sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -562,8 +593,12 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
   // FloatConverter
   private[this] def castToFloat(from: DataType): Any => Any = from match {
     case StringType =>
-      buildCast[UTF8String](_, s => try s.toString.toFloat catch {
-        case _: NumberFormatException => null
+      buildCast[UTF8String](_, s => {
+        val floatStr = s.toString
+        try floatStr.toFloat catch {
+          case _: NumberFormatException =>
```

Comment:
@viirya So my idea here was that.. since we don't expect these kind of literals a lot i.e its not a common case .. we don't change our normal processing path to add any possible runtime costs. Thats why we keep all these in the exception processing.

Reply:
I like the @viirya approach for readability and could you check actual performance changes?

Reply:
@maropu Oops.. didn't see this comment. I suppose i have to use the benchmark framework for this ? Appreciate any tip on this..

Reply:
Could we check the numbers by a simple query?, e.g.,
```
scala> spark.range(10).selectExpr("CAST(double(id) AS STRING) a").write.save("/tmp/test")
scala> spark.read.load("/tmp/test").selectExpr("CAST(a AS DOUBLE)").write.format("noop").save()
```
In another pr, I observed that a logic depending on exceptions cause high performance penalties: https://github.com/lz4/lz4-java/pull/143
So, I'm a bit worried that this current logic has the same issue.

Reply:
@maropu Ok.. i will give it a try. One thing to note here is that, we didn't introduce a try/catch in this PR. That was existing before. We just added some extra code in the catch block. However, i will give it a try and get back.

Reply:
Yea, I mean the performance for known strings as @viirya said.

Reply:
@maropu I did the comparison. So our current implementation of handling these special values in the catch block seems to perform better than putting these in the main code path.  Here are the results. I didn't make the change to the codegen path. So lets ignore the second line. The first set of numbers are for the run with current code path. 2nd set of numbers are when these special values are handled in the main code path.
```
Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.14.6
[info] Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
[info] cast from string to double:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] cast from string to double wholestage off           5076           5394         450          0.0       50759.0       1.0X
[info] cast from string to double wholestage on           4709           4876         194          0.0       47093.1       1.1X
[info]


Java HotSpot(TM) 64-Bit Server VM 1.8.0_202-b08 on Mac OS X 10.14.6
[info] Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
[info] cast from string to double:               Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] cast from string to double wholestage off           5002           5040          53          0.0       50023.6       1.0X
[info] cast from string to double wholestage on           4926           4976          45          0.0       49258.8       1.0X
~                                                                                                                                 
```

Please let me know.

Reply:
Many thanks! I see. What was input data? Some parts of input data include `inf`/`-inf`?

Reply:
@maropu Actually my input data didn't contain any of these special literals. Basically i was testing with the condition when we don't hit the catch block. Basically we are trying optimize for the best case ?

Reply:
I just make sure that a small number of `inf/-inf` does not worse the prformance. If this is true, I think the current code looks ok to me.

Reply:
@maropu In the mean time i had tried with 1% of data being 'inf' and i can confirm that it does not hurt the performance :-)

Reply:
Sounds good.

Reply:
Sounds nice, too ;)